src/services/filesync/filesync.ts

```diff
@@ -325,97 +322,117 @@ export class FileSync {
   }
 
   /**
-   * Synchronizes local files with Gadget files. If there are local
-   * changes, prompts the user to either merge or reset them.
+   * Ensures the local filesystem is in sync with Gadget's filesystem.
+   * - All non-conflicting changes are automatically merged.
+   * - Conflicts are resolved by prompting the user to either keep their
+   *   local changes or keep Gadget's changes.
+   * - This function will not return until the filesystem is in sync.
    */
-  async sync(): Promise<void> {
-    const data = await this.editGraphQL.query({ query: REMOTE_FILES_VERSION_QUERY });
-    const remoteFilesVersion = BigInt(data.remoteFilesVersion);
-    const hasRemoteChanges = remoteFilesVersion > this.filesVersion;
-
-    const getLocalChanges = async (): Promise<Changes> => {
-      const changes = new Changes();
-      for await (const normalizedPath of this.directory.walk()) {
-        if (normalizedPath.startsWith(".gadget/")) {
-          // ignore changes to the .gadget/ directory
-          continue;
-        }
+  async sync({ attempt = 0 }: { attempt?: number } = {}): Promise<void> {
+    if (attempt > 10) {
+      throw new TooManySyncAttemptsError(attempt);
+    }
+
+    const { filesVersionHashes, localHashes, gadgetHashes, gadgetFilesVersion } = await this._getHashes();
```

Comment:
Am I getting this right? If you start a sync and there is 1 file that has been modified we will sync the project server side 4 times?

- Attempt = 0, filesVersionHashes
- Attempt = 0, gadgetFilesVersionHashes
- Attempt = 1, filesVersionHashes
- Attempt = 1, gadgetFilesVersionHashes

Because we always iterate through this at least once if there has been any change either locally or upstream?

Reply:
Yes that's right.